### PR TITLE
refactors sending emails to admins when a user joins a loop

### PR DESCRIPTION
**Pull request by gclamigueiro**
This PR introduces the following changes:
1 - Creation of the Email Service with the following functions:

1.1 - **EmailLoopAdminsOnUserJoin**: Given a User and a list of chainIds, it sends an email to every admin of those loops.

1.2 - **EmailYouSignedUpForLoop**: Given a User and a list of chainNames, it sends a confirmation email to the User for every loop.

2 - Replacing the email-sending logic in the chain and login controllers with a service call.

Some questions I have related to the changes:

I had to remove the **go** keyword because I encountered an error in my local environment. Is there a way to check if there is a problem with my local setup so I can add it again? If it's a general problem, then I could look for another way to send the email asynchronously.
 
Related Issue: https://github.com/the-clothing-loop/website/issues/601

### CODE REVIEW

**Comment by lil5** [comment]
@gclamigueiro have you tested this? Is it ready for acceptance?
Leave the async out for now 👍

If so I'd like to get this in acceptance _asap_ would @NatalieWilson19 and @tobifaf be able to take a look?

**Comment by gclamigueiro** [comment]
> @gclamigueiro have you tested this? Is it ready for acceptance? Leave the async out for now 👍
> 
> If so I'd like to get this in acceptance _asap_ would @NatalieWilson19 and @tobifaf be able to take a look?

Hi Lucian, yes I tested it. I tested the scenario when you join a loop by registering, and when you join by clicking "join loop" button and you are already logged in.  
However, I was not able to test the case when the user had multiple loops to join. I think that is not possible anymore because if you try to join another loop without validating your email first the webpage redirects you to register again. 

nevertheless, more tests are welcome in case I missed some edge case.

**Comment by NatalieWilson19** [comment]
I pulled down these changes and it works as expected when the user signs up for a loop, both when the user is logged in and logged out. I also tested for when a loop has one admin or multiple and it works as expected for both those cases 👍

Also, I agree with Gabriel that I don’t think it's possible for the user to sign up for more than one loop at a time, since the website redirects you to login if you try to sign up for another loop before verifying your email

File: server/internal/services/email.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	}
+
+	if len(results) == 0 {
+		goscope.Log.Errorf("Empty chain that is still public: ChainID: %d", chainIds)
+		return fmt.Errorf("No admins exist for this loop")
+	}
+
+	for _, result := range results {
+		if result.Email.Valid {
+			views.EmailSomeoneIsInterestedInJoiningYourLoop(db, result.I18n,
```
Comment (tobifaf) [inline server/internal/services/email.go:33]: The returned error of this function is not handled as of now.

File: server/internal/services/email.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+			)
+		}
+	}
+
+	return nil
+}
+
+func EmailYouSignedUpForLoop(db *gorm.DB, user *models.User, chainNames ...string) error {
+	for _, chainName := range chainNames {
+		views.EmailYouSignedUpForLoop(db, user.I18n, user.Name, user.Email.String, chainName)
```
Comment (tobifaf) [inline server/internal/services/email.go:51]: Also error not handled

File: server/internal/controllers/chain.go [hunk 3]
```diff
@@ -379,34 +380,13 @@ LIMIT 1
...
+		err := services.EmailLoopAdminsOnUserJoin(db, user, chain.ID)
 		if err != nil {
-			goscope.Log.Errorf("Error retrieving chain admins: %s", err)
-			c.String(http.StatusInternalServerError, "No admins exist for this loop")
-			return
-		}
-		if len(results) == 0 {
-			goscope.Log.Errorf("Empty chain that is still public: ChainID: %d", chain.ID)
-			c.String(http.StatusInternalServerError, "No admins exist for this loop")
+			goscope.Log.Errorf("Unable to send email to associated loop admins: %v", err)
```
Comment (tobifaf) [inline server/internal/controllers/chain.go:385]: if the 500 is raised here, the user never receives a message, but the DB is already updated. And if they return, the signup is "waiting for approval". Maybe it's better to at least send the email to the user in all cases?

File: server/internal/controllers/chain.go [hunk 4]
```diff
@@ -379,34 +380,13 @@ LIMIT 1
...
-					chain.Name,
-					user.Name,
-					user.Email.String,
-					user.PhoneNumber,
-					user.Address,
-					user.Sizes,
-				)
-			}
-		}
+		services.EmailYouSignedUpForLoop(db, user, chain.Name)
```
Comment (tobifaf) [inline server/internal/controllers/chain.go:389]: also no error handled here.

**Comment by tobifaf** [comment]
can also verify everything works and the refactor looks like a great improvement @gclamigueiro . @lil5 and @gclamigueiro i added some remarks about the error handling, but not sure if this if/should be in scope. Although I think the order of DB insert -> email hosts -> email user could be reconsidered. Or perhaps a DB rollback in case of an error?

File: server/internal/controllers/chain.go (hunk 4 above)
Comment (tobifaf) [reply server/internal/controllers/chain.go:389]: also: this function does not seem to be complete: the templates for "you_signed_up_for_loop" are missing?

**Comment by lil5** [comment]
Merging we'll fix the error handling afterwords

**Comment by gclamigueiro** [comment]
> can also verify everything works and the refactor looks like a great improvement @gclamigueiro . @lil5 and @gclamigueiro i added some remarks about the error handling, but not sure if this if/should be in scope. Although I think the order of DB insert -> email hosts -> email user could be reconsidered. Or perhaps a DB rollback in case of an error?


Hi @tobifaf. Yes, I can add the code for logging errors. However, in this particular case, in my opinion, it is not recommended to perform a rollback. For instance, if the user is attempting to validate their email, and the email sending process fails, and we ask them to verify again, but the issue persists, the user can't validate their email, which is an important part of the flow. In my opinion, it is better that the user can continue with the normal flow.

Additionally, if I am not mistaken, there is a retry mechanism for failed emails. If this scenario is not covered by the existing mechanism, it may be better to incorporate it into the system and not burden the user with the task of retrying.
